lib/collision_detector_2.cpp

```diff
@@ -0,0 +1,276 @@
+#include "brickengine/components/physics_component.hpp"
+#include "brickengine/std/sign.hpp"
+#include <cmath>
+#include <algorithm>
+
+
+CollisionDetector2::CollisionDetector2(std::unordered_map<std::string, std::set<std::string>> is_trigger_tag_exceptions,
+                                       EntityManager& em) : em(em), is_trigger_tag_exceptions(is_trigger_tag_exceptions) {}
+
+std::vector<Collision> CollisionDetector2::detectCollision(int entity_id) {
+    auto physics = em.getComponent<PhysicsComponent>(entity_id);
+    std::vector<Collision> collisions;
+    if (physics->collision_detection == CollisionDetectionType::Discrete) {
+        auto discrete_collisions = this->detectDiscreteCollision(entity_id);
+        std::for_each(discrete_collisions.begin(), discrete_collisions.end(), [&collisions](DiscreteCollision& dc) {
+            collisions.push_back(Collision(dc.opposite_id, dc.is_trigger));
+        });
+    } else if (physics->collision_detection == CollisionDetectionType::Continuous) {
+        {
+            auto collision = this->detectContinuousCollision(entity_id, Axis::X, Direction::NEGATIVE);
+            if (collision.space_left >= 0 && collision.opposite_id) {
+                collisions.push_back(Collision(*collision.opposite_id, false));
+            }
+        }
+        {
+            auto collision = this->detectContinuousCollision(entity_id, Axis::X, Direction::POSITIVE);
+            if (collision.space_left <= 0 && collision.opposite_id) {
+                collisions.push_back(Collision(*collision.opposite_id, false));
+            }
+        }
+        {
+            auto collision = this->detectContinuousCollision(entity_id, Axis::Y, Direction::NEGATIVE);
+            if (collision.space_left >= 0 && collision.opposite_id) {
+                collisions.push_back(Collision(*collision.opposite_id, false));
+            }
+        }
+        {
+            auto collision = this->detectContinuousCollision(entity_id, Axis::Y, Direction::POSITIVE);
+            if (collision.space_left <= 0 && collision.opposite_id) {
+                collisions.push_back(Collision(*collision.opposite_id, false));
+            }
+        }
+    }
+    return collisions;
+}
+
+bool CollisionDetector2::findDisplacementException(std::set<std::string> tags_1, std::set<std::string> tags_2) const {
+    for (const std::string& tag_1 : tags_1) {
+        if (is_trigger_tag_exceptions.count(tag_1)) {
+            for (const std::string& tag_2 : tags_2) {
+                if (is_trigger_tag_exceptions.at(tag_1).count(tag_2)) {
+                    return true;
+                }
+            }
+        }
+    }
+    for (const std::string& tag_2 : tags_2) {
+        if (is_trigger_tag_exceptions.count(tag_2)) {
+            for (const std::string& tag_1 : tags_1) {
+                if (is_trigger_tag_exceptions.at(tag_2).count(tag_1)) {
+                    return true;
+                }
+            }
+        }
+    }
+    return false;
+}
```

Comment:
I would add some comments about the functionality of this method. I do not know what it does, as it is called 'find', but it returns a boolean when looking for exceptions at two tags?